### ISSUE TITLE
CS8629: False positive with Nullable<T>.Value after proving that its .HasValue matches another instance's .HasValue, which is proven true

### ISSUE THREAD

**Issue by airbreather** [dotnet/roslyn#53115]
**Version Used**: 
Repros on SharpLab on the latest commit to `main` (bc98ed9) as well as Visual Studio Enterprise 16.9.4.
**Steps to Reproduce**:
Compile the following:
```csharp
#nullable enable
using System;
static double Sum(double? a, double? b)
{
    if (a.HasValue != b.HasValue)
    {
        throw new ArgumentException("Both must be null or non-null together.");
    }

    return a.HasValue
        ? a.Value + b.Value
        : double.NaN;
}
```
SharpLab snippet: https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgABAJgEYzzqA2XAEwHtks8BlZMACjYccAflwQANC3adRwAJRkA3mVyrcUAGa4+EAHQAJCAGcAahFTI8AQgC8uYAeNmL2BaTW5l7j2oAuACwAnVgB3XBhsMIBBQIBzXgJfAFEADwBjbAAHXyhWGD4AIgAhVgDcMGQjX3s8FHRcVkDwvIBaOtRcX1ZY7ADsQN0CuQBuFTUAXzIx1XIAdjFHU3NLaZ9RPWdLXABqe11N7FWPECkhbF0AOQgL0dJxoA=

**Expected Behavior**:
No nullability warnings

**Actual Behavior**:
CS8629: Nullable value type may be null.

Specifically, it warns that `b.Value` is not safe, even though it provably is:
1. `b.HasValue` is equal to `a.HasValue`
1. `a.HasValue` is `true`
1. Therefore, `b.HasValue` is `true`
1. Therefore, `b.Value` is safe

**Workarounds**:
Of course, I can always just suppress, change it to `.GetValueOrDefault()`, or rewrite it so that the two conditions are combined, to something like:
```csharp
#nullable enable
using System;
static double Sum(double? a, double? b)
{
    if (a.HasValue)
    {
        if (!b.HasValue)
        {
            throw new ArgumentException("Both must be null or non-null together.");
        }

        return a.Value + b.Value;
    }
    else if (b.HasValue)
    {
        throw new ArgumentException("Both must be null or non-null together.");
    }
    else
    {
        return double.NaN;
    }
}
```

What I want to avoid is something like this, however, as it adds a branch that I don't think can be covered (so code coverage looks worse):
```csharp
#nullable enable
using System;
static double Sum(double? a, double? b)
{
    if (a.HasValue != b.HasValue)
    {
        throw new ArgumentException("Both must be null or non-null together.");
    }

    return a.HasValue && b.HasValue
        ? a.Value + b.Value
        : double.NaN;
}
```

And changing `&&` to `&` is a bit too esoteric.

**Comment by RikkiGibson** [comment]
This seems to be analogous to #36927. See https://github.com/dotnet/roslyn/issues/36927#issuecomment-508595947 for why it's unlikely our analysis will learn to do the right thing in this situation.

You can get the desired nullable analysis while reducing the number of tests performed in the implementation using patterns in a switch expression or switch statement. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgyABAJgEYzKaA2XAEwHtks8BlZMACnaccAflwQANKw5cxwAJRkA3mVxrcFAOy5+k3AtwBnAO5QALgGMAFqvUrS6xzqFdxANQipk2KS5z6PL2x5XABeAD53T29cAGoA6J9bJx0UdCk01BCI6WFsADoAOQhCiWSnAH0wyLMrACc2Y1wYbCaAQTqAcz4CMwBRAA8LbAAHMyg2GH4AIgAhNlrcMGRDM308TNw2OubJgFpNszZO7FrsOvzpxQd1AF8AbjJbsiA===)

```cs
#nullable enable

using System;

static double Sum(double? a, double? b)
{
    return (a, b) switch
    {
        (double aValue, double bValue) => aValue + bValue,
        (null, null) => double.NaN,
        _ => throw new ArgumentException("Both must be null or non-null together.")
    };
}
```

Alternatively you can use an assertion to silence the warning:

```cs
#nullable enable
using System;
using System.Diagnostics;

static double Sum(double? a, double? b)
{
    if (a.HasValue != b.HasValue)
    {
        throw new ArgumentException("Both must be null or non-null together.");
    }

    if (a.HasValue)
    {
        Debug.Assert(b.HasValue);
        return a.Value + b.Value;
    }
    
    return double.NaN;
}
```
